add identitymanager package

ianvernon opened the pull request:
This package will be used to track what global identities are being consumed by the currently running cilium-agent, i.e. what identities are being used by endpoints that are currently running.

This will be used to track what identities are being consumed by endpoints on the local node in the effort to decouple policy generation from being per-endpoint to being per-identity. 

The daemon hooks into this package at the following points:

* Add / increment reference count to identity in the identitymanager when an endpoint has its SecurityIdentity field set to a specific identity
* Delete / decrement reference count to an old identity for an endpoint from identitymanager when an endpoint has an identity change.
* Delete / decrement reference count to an old identity for an endpoint from identitymanager when an endpoint is deleted.

A new CLI option is provided to `cilium identity list`, with the `--endpoints` argument to allow for introspection into the current state of the identitymanager by the cilium-agent:

Example:
    
```
    $ cilium identity list --endpoints
    ID      LABELS                                                 REFCOUNT
    4       reserved:health                                        1
    104     k8s:io.cilium.k8s.policy.cluster=default               1
            k8s:io.cilium.k8s.policy.serviceaccount=coredns
            k8s:io.kubernetes.pod.namespace=kube-system
            k8s:k8s-app=kube-dns
    24857   k8s:appSecond=true                                     2
            k8s:id=app2
            k8s:io.cilium.k8s.policy.cluster=default
            k8s:io.cilium.k8s.policy.serviceaccount=app2-account
            k8s:io.kubernetes.pod.namespace=default
            k8s:zgroup=testapp
    33585   k8s:id=app1                                            4
            k8s:io.cilium.k8s.policy.cluster=default
            k8s:io.cilium.k8s.policy.serviceaccount=app1-account
            k8s:io.kubernetes.pod.namespace=default
            k8s:zgroup=testapp
    58550   k8s:id=app3                                            2
            k8s:io.cilium.k8s.policy.cluster=default
            k8s:io.cilium.k8s.policy.serviceaccount=default
            k8s:io.kubernetes.pod.namespace=default
            k8s:zgroup=testapp
```
    
This shows that for identity 33585, 4 endpoints are consuming this identity.

Notable implementation details:

* The map is keyed by the numeric identity, not a pointer to an identity. The same identity can have multiple `identity.Identity` instances on the local node, so using the pointer as a key in the map is not feasible.

I am happy to add a runtime test once I get sign-off on the API side of things. I've tested it locally.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7616)
<!-- Reviewable:end -->

Commit:
add identitymanager package

This package will be used to track what global identities are being consumed
by the currently running cilium-agent, i.e. what identities are being used
by endpoints that are currently running. When no endpoint references an
identity in the identitymanager any more, it is removed from the
identitymanager.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: update global identitymanager when identity changes

Add an identity into the identitymanager when an endpoint has its
SecurityIdentity field set to a specific identity, and delete its old identity
as well.

Also delete the endpoint's identity from the identitymanager when the endpoint
is deleted.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/identity/identitymanager/manager.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+// Delete deletes the identity from the GlobalIdentityManager.
+func Delete(identity *identity.Identity) {
```
I don't think we want an `Upsert()`/`Delete()` here. If you do that, then there's no point for the reference tracking; the caller then has to track references to figure out when to call `Delete()`. I'll write this feedback up in the main feedback to this PR, but I think we should be doing `Add`/`Remove` semantics instead of `Upsert()`/`Delete()`.

ianvernon replied:
To clarify, you're saying that I should change the names of the functions based on semantics as opposed to the underlying implementation?

joestringer replied:
I was saying that both the implementation and naming should change to `Add()`/`Remove()`.

Semantically, if you use `Upsert()`, then you are telling the caller that they don't have to care about whether they insert the same identity twice. However, if the underlying implementation is doing reference tracking, then you want to enforce that the caller is appropriately adding and removing references to the identity for the duration that they are using the value. That way, this manager can know when no-one references the identity any more and can take action (eg, stop tracking the references and issue some event).

Alternatively, if you're looking for `Upsert()` and `Delete()` semantics, then I don't see the point in having the implementation track the references inside. If something else has the knowledge to issue a `Delete()` because it knows that no-one else is still using the identity, then the references aren't providing any value.

ianvernon replied:
I understand the distinction between changing it to `Add` and `Remove`, but I don't understand what you want me to change on the implementation side.

The `identitymanager` isn't aware of what is consuming the identities within it, just that something elsewhere is consuming the identities. 

Are you saying that we should also track what is consuming the identities within `identityManager`? E.g., a list of pointers to some interface, e.g., `Endpoint`, where `Endpoint` has a single function `GetSecurityIdentity`? This would ensure that we wouldn't be able to `Delete` if the consumer of the identity did not previously add it, as we would pass a pointer to the Endpoint to the `Add` and `Remove` functionality to indicate ownership.

I'm unsure how to otherwise enforce that the caller (the endpoint lifecycle) is appropriately adding and removing references to the identity for the duration that it is using the value apart from some type of integration test (which I am planning on adding once I get sign-off on the API).

joestringer commented on pkg/identity/cache/allocator.go:
```diff
@@ -227,6 +228,10 @@ func Release(ctx context.Context, id *identity.Identity) (bool, error) {
 
+	// identitymanager only tracks global identities in use by local endpoints,
+	// not local identities.
+	identitymanager.Delete(id)
```
I don't think this holds the intent in the way that we want it to.

AFAIK this `Release()` function here is only called when the global kvstore deletes all references to this Identity. So, if you schedule an endpoint onto this node and track the Identity, then delete the endpoint from this node, so long as there is still an instance of that identity somewhere in the cluster, the identity reference will be retained. This doesn't quite match up with the intent of tracking local endpoint identities.

/cc @jrajahalme

joestringer replied:
As discussed offline, I could be completely off the wall on this but there's definitely a take-away here about "how are we testing/validating this behaviour?", in particular where the same identity is used by endpoints on multiple nodes.

ianvernon replied:
I've changed this behavior, it should be OK now.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -552,6 +553,7 @@ func (e *Endpoint) SetIdentity(identity *identityPkg.Identity) {
 		oldIdentity = e.SecurityIdentity.StringID()
 	}
 
+	identitymanager.Upsert(identity)
```
This is fine, I think we should do an `identitymanager.Remove(oldIdentity)` here as well.

EDIT: OK, well the current `oldIdentity` variable is a string, you'll need to turn that into a real identity reference (or nil) and handle a bit better. Still, only a minor refactor.

ianvernon replied:
LMK what you think of the new behavior.

joestringer reviewed (changes requested):
Thanks for jumping on this so fast!

I think the semantics of `Add()` / `Remove()` are more appropriate than `Upsert()`/`Delete()`. The idea behind this is that the endpointmanager should `Add()` a reference to an identity whenever a new endpoint shows up with the particular Identity, and `Remove()` a reference whenever the endpoint is removed. This means that if two endpoints with the same Identity show up and then one is removed, the identity manager will retain an Identity reference for the remaining endpoint.

With the PR as-is, it appears that the real reference counting for the Identity is happening in the kvstore and we're relying on the kvstore deletion to remove the identity for this cache. But then this cache isn't tracking local endpoint identities any more, it's tracking identities that have at some point been used locally. If you deploy a couple of instances of an app into a large cluster, then redeploy the app again, it will likely be scheduled onto other nodes in the cluster but the identitymanager on the initial nodes will not clean up references to those identities.

Here's the points I'd expect there to be `Add()`/`Remove()`:
* When adding an endpoint. This could perhaps be from endpointmanager, or perhaps somewhere inside pkg/endpoint, depending on where there's a point where the logic converges. Ideally after all of the error cases are handled so that we can avoid needing to release in all the error paths. If not possible, be sure to remove the reference in those error paths.
* When removing an endpoint. Again, could be endpointmanager or could be in the endpoint's LeaveLocked() function.
* When changing the labels
  * Looks this PR is half way there at the moment, just need to `Remove()` a reference to the old identity.

Commit:
api: add identity/endpoints api

This API returns the set of identities which are being used by endpoints for
the local instance of cilium, as opposed to all identities in the cluster.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
daemon: handle identity/endpoints API

The daemon can now handle the api/v1/identity/endpoints API, which will return
the list of all identities that are in-use by endpoints in the
GlobalIdentityManager.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
cmd: add `cilium identity list --endpoints` command

This command allows for introspecting about the current state of the
GlobalIdentityManager in cilium.

Example:

```
$ cilium identity list --endpoints
getting identities endpoints
ID      LABELS                                                 REFCOUNT
4       reserved:health                                        1
104     k8s:io.cilium.k8s.policy.cluster=default               1
        k8s:io.cilium.k8s.policy.serviceaccount=coredns
        k8s:io.kubernetes.pod.namespace=kube-system
        k8s:k8s-app=kube-dns
24857   k8s:appSecond=true                                     2
        k8s:id=app2
        k8s:io.cilium.k8s.policy.cluster=default
        k8s:io.cilium.k8s.policy.serviceaccount=app2-account
        k8s:io.kubernetes.pod.namespace=default
        k8s:zgroup=testapp
33585   k8s:id=app1                                            4
        k8s:io.cilium.k8s.policy.cluster=default
        k8s:io.cilium.k8s.policy.serviceaccount=app1-account
        k8s:io.kubernetes.pod.namespace=default
        k8s:zgroup=testapp
58550   k8s:id=app3                                            2
        k8s:io.cilium.k8s.policy.cluster=default
        k8s:io.cilium.k8s.policy.serviceaccount=default
        k8s:io.kubernetes.pod.namespace=default
        k8s:zgroup=testapp
```

This shows that for identity 33585, 4 endpoints are consuming this identity.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on api/v1/models/cluster_status.go:
```diff
@@ -16,7 +16,6 @@ import (
 
 // ClusterStatus Status of cluster
 // swagger:model ClusterStatus
-// +k8s:deepcopy-gen=true
```
Is this intentional? (Please scan the PR for similar occurrences, I see a few)

ianvernon replied:
All I did was run `make generate-api` locally, which AFAIK is the correct way to regenerate the API per the documentation.

Looks like 6928b56bc6 edited the files directly =/ . I'll remove the changes to these files.

ianvernon replied:
Fixed - this is going to be obnoxious for future API work, though. We will have to manually undo changes to code which we need `deepcopy-gen` functionality for :(

joestringer replied:
Yep, I for one find this pretty confusing. Not sure whether someone has looked into any changes we might make to the tooling to prevent this from happening.

raybejjani replied:
I tried to sort it out (I promised aanm I would). Unfortunately, we can't make a separate file that lists which types need a k8s deepcopy generated (one of the other autogenenarated thing can take an annotation on the type, or a list somewhere else). The only thing I can think of is to have a separate script/tool that takes a list of types and inserts the `// +k8s:deepcopy-gen=true` line.

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -356,6 +356,30 @@ paths:
+          description: Invalid identity format in storage
+          x-go-name: InvalidStorageFormat
+          schema:
+            "$ref": "#/definitions/Error"
```
I see these 520/521 in the identity side as well, but I don't see them being used anywhere in the codebase. Do we need them? When do we envision sending such responses back?

ianvernon replied:
Probably not - we can remove them for now.

ianvernon replied:
Removed.

joestringer reviewed (changes requested):
The new implementation makes sense to me. The naming still confuses me though (see my response to the prior thread). If my earlier response doesn't make any sense, feel free to reach out on slack or somewhere and we can discuss more.

I see some potentially unintended changes below, and also some unused API error codes.

raybejjani commented on pkg/identity/identitymanager/manager.go:
```diff
@@ -0,0 +1,160 @@
+	// for easy updating / tracking lifecycles of identities on the local node
+	// without having to pass around a specific instance of an IdentityManager
+	// throughout Cilium.
+	GlobalIdentityManager = NewIdentityManager()
```
We don't really use `GlobalIdentityManager` anywhere but this package. Was there a scenario where you expected to use GlobalIdentityManager or the `IdentityManager` type again? 
Is the reason we have the package level functions to mimic endpoint manager?

I guess when I saw `GlobalIdentityManager `, I was expecting explicit references to it at the places we use it.

ianvernon replied:
I wanted to be able to create multiple instances of the `IdentityManager` type for unit testing, as well as have a global singleton for direct reference from endpoint so I wouldn't have to pass a reference to `Owner` around.

raybejjani commented on pkg/identity/identitymanager/manager.go:
```diff
@@ -0,0 +1,160 @@
+func (idm *IdentityManager) Add(identity *identity.Identity) {
+	idm.mutex.Lock()
+	defer idm.mutex.Unlock()
+	idm.add(identity)
```
Calling `RemoveOldAddNew(nil, identity)` might make it more clear that Add, Remove and RemoveOldAddNew all use the same code-paths. I'm not sure that it is ambiguous, though. shrug.

raybejjani commented on pkg/identity/identitymanager/manager.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+// GetIdentityModels returns the API representation of the IdentityManager.
+func (idm *IdentityManager) GetIdentityModels() []*models.IdentityEndpoints {
```
Probably not worth changing anything, but `models.IdentityEndpoints` isn't defined when this commit is made.

raybejjani commented on pkg/identity/identitymanager/manager_test.go:
```diff
@@ -0,0 +1,90 @@
+	c.Assert(exists, Equals, false)
+
+	_, exists = idm.identities[barIdentity.ID]
+	c.Assert(exists, Equals, true)
```
Should this also assert that the refcount is still 1 for `barIdentity`?

raybejjani commented on pkg/identity/identitymanager/manager_test.go:
```diff
@@ -0,0 +1,90 @@
+	c.Assert(exists, Equals, false)
+
+	idm.Add(fooIdentity)
+	c.Assert(idm.identities[fooIdentity.ID].refCount, Equals, uint(1))
```
I was going to suggest creating a `GetRefcount(id Identity) uint` function and test with that, but it is basically this line of code. One thing I'm not sure about is that the only way we export the refcount is via `GetIdentityModels` but we aren't testing that here (and it would be a pain to assert on its members).

raybejjani commented on Documentation/cmdref/cilium_identity_list.md:
````diff
@@ -15,6 +15,7 @@ cilium identity list [LABELS] [flags]
 ### Options
 
 ```
+      --endpoints       list identities of locally managed endpoints
````
I was a little surprised that this listed the refcount and not the actual endpoints using each identity. I don't know what to name this instead, though, apart from `used-locally` or something equally obtuse. Do we ever anticipate adding a column with the endpoint IDs that have each identity?

ianvernon replied:
I'm open to changing the flag. `--used-locally` sounds fine to me. I can change that in a follow-up.

joestringer replied:
I had the same initial thought. That would be a nice extension to this.

raybejjani reviewed (approved):
My only comments are effectively formatting changes. This can be merged as-is.
